text/3720-labeled-match.md

```diff
@@ -0,0 +1,1419 @@
+- the (expert) programmer definitely wants the desugaring into a `goto`
+
+So, `loop match` is a solid way to make progress on better codegen. Improved optimizations on MIR are also very welcome, but never entirely remove the need for `loop match` from a programmer's perspective.
+
+## recognize "loop + match" and optimize
+
+In theory it is possible to internally recognize and rewrite a "loop + match" expression into a `loop match`. With this approach, no changes to language syntax are needed.
+
+A fundamental problem with this approach is a change in drop order:
+
+```rust
+let mut state = 0;
+'label: loop {
+    match state {
+        0 => {
+            let x = vec![1,2,3];
+            state = 1;
+            // drop of `x` gets inserted between state update and jump
+            continue 'label;
+        }
+        _ => ...
+    }
+}
+
+// versus if you rewrite to `loop match`
+
+loop match 0 {
+    0 => {
+        let x = vec![1,2,3];
+        // drop of `x` happens before the state update
+        continue 1;
+    }
+    _ => ...
+}
+```
+
+Beyond that, the analysis for recognizing "loop + match" will likely be complex and fragile. Part of the appeal of `loop match` is that the desugaring rules are simple and deterministic. Using a `loop match` is intentional, and signals that something subtle is going on: for readers and future reviewers it is clear that the `loop match` desugaring is desired and potentially crucial for the code to perform well.
+
+## introduce just `continue <operand>`
+
+Do we really need `loop match`? We could instead allow `continue <operand>` inside of `loop { match _ { ... } }`.
+
+An advantage of this approach is that one can define macros that have the loop's label in scope:
+
+```rust
+let mut state = 0;
+
+'label: loop {
+    macro_rules! foo {
+        () => {
+            continue 'label (state + 1)
+        }
+    }
+
+    match state {
+        1 => foo!(),
+        2 => foo!(),
+        _ => todo!(),
+    }
+}
+```
+
+The main downside here is that it is really subtle that the `continue <operand>` expression updates the match scrutinee in the next iteration.
+
+## don't introduce irreducible control flow
+
+It seems possible to delay the desugaring of a `const`-annotated state transition to a goto until after borrow checking. In that case we'd annotate the relevant `goto`s to the top of the loop with a "shortcut", the location where they will actually jump to.
+
+```
+    bb1: {
+        _3 = discriminant(_2);
+        switchInt(move _3) -> [0: bb4, 1: bb3, otherwise: bb2];
+    }
+
+    bb4: {
+        _2 = const State::B;
+        goto -> bb1; // <- would be annotated with "please make this a `goto -> bb3` after borrow checking"
+    }
+```
+
+The advantages of having more accurate borrow checking, and accepting more valid programs, are compelling to me, but this more limited solution here could absolutely work from just a performance perspective.
+
+## Syntax Squables
```

Comment:
Could you still put https://github.com/rust-lang/rfcs/pull/3720#issuecomment-2443280361 / https://github.com/rust-lang/rfcs/pull/3720#issuecomment-2451823178 as an alternative? Basically using a dedicated macro rather than introducing a new syntax for this feature in the first place, so we can decide later a dedicated syntax and whether it is worth it (like `addr_of!(x)` → `&raw x` or `matches!(x, p)` ↛ `x is p`).

```rust
finite_state_machine! {
    goto dyn count % 4;
    0 => {
        one();
        goto 3;
    }
    3 => {
        one();
        goto 2;
    }
    2 => {
        one();
        goto 1;
    }
    1 => {
        one();
        n -= 1;
        if n > 0 {
            goto 0;
        } else {
            break;
        }
    }
    _ => unreachable!(),
};
```

Reply:
yes, I added it. 

I quite like `goto dyn` actually, that's good.

I am, at first glance, worried about the ergonomics of such a macro. To get the `goto` target validation and lowering to a MIR goto you need serious compiler support. You'd also need custom error messages for this special rust-like dialect that is valid within the macro, and it needs to be parsed. Maybe there is actually good tooling for this, but my experience with the `asm!` macros is that the parsing is very manual and kind of tedious.